### PR TITLE
Move TokenBundleSizeAssessor

### PR DESCRIPTION
**Pull request by Anviking**
- [x] Move `TokenBundleSizeAssessor` to new `Cardano.Wallet.Write.Tx.Balance.TokenBundleSizeAssessor` module.
    - I left a version of the `TokenBundleMaxSize` newtype in `Primitive.Types`, as it is entangled with one remaining thing: the protocol parameters API endpoint.

### Issue Number

ADP-3081

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -1216,7 +1216,6 @@ data ApiNetworkParameters = ApiNetworkParameters
     , activeSlotCoefficient :: !(Quantity "percent" Double)
     , decentralizationLevel :: !(Quantity "percent" Percentage)
     , desiredPoolNumber :: !Word16
-    , maximumTokenBundleSize :: !(Quantity "byte" Natural)
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: It's tempting to remove this field, and I think it should be fine, but requires a discussion

Comment (Anviking) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: We could also leave the field, but remove the dependency on `balance-transaction:TokenBundleMaxSize`, e.g. by using `TxSize` instead. This would achieve decoupling, and defer the questioning of the worth of the `maximumTokenBundleSize` wallet api pparam field for later.

Comment (Anviking) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: @jonathanknowles thoughts?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 2]
```diff
@@ -2098,6 +2035,42 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
                     ]
             tx `shouldBe` Left ErrBalanceTxMaxSizeLimitExceeded
 
+    describe "effect of maxValSize" $ do
+        it "change split when needed" $ property $ \(room :: Small Int) -> do
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: > we could also imagine dropping most of it, and adding a few goldens on the balanceTx / black-box level.

Ah right, I did add this test here in this PR

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs (hunk 1 above)
Comment (Anviking) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: I've now kept `maximumTokenBundleSize` and duplicated the `TokenBundleMaxSize` definition.

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs [hunk 3]
```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | Assessing sizes of token bundles
+module Cardano.Wallet.Write.Tx.Balance.TokenBundleSizeAssessor
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs]: Can we shorten the name to `TokenBundleSize`?

Justification:
- It's more concise
- This module includes a greater variety of things than just the `TokenBundleSizeAssessor` type. For example:
    - `TokenBundleMaxSize`: defines the maximum size of a token bundle
    - `computeTokenBundleSerializedLengthBytes`: computes the serialised length of a token bundle

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+          [ fromIntegral . max 0 . (4000 +) <$> arbitrary @Int
+
+          -- Generate more extreme values (both small and large)
+          , fromIntegral <$> arbitrary @Word64
+          ]
+    shrink (TokenBundleMaxSize (TxSize s)) =
+        map (TokenBundleMaxSize . TxSize . fromIntegral)
+        . shrink @Word64 -- Safe w.r.t the generator, despite TxSize wrapping a
+                         -- Natural
+        $ fromIntegral s
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs]: Shouldn't this `Arbitrary` instance be defined within the `Spec` module?

(If for some reason we need to include this logic within the implementation, then can we follow our existing pattern and define `{gen,shrink}TokenBundleMaxSize` inside a separate `Gen` module? 🙏🏻)

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs]:
```suggestion

```

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    ( ppMaxValSizeL )
+import Control.Lens
+    ( (^.) )
+
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Data.ByteString as BS
+
+-- | The maximum size of a serialized `TokenBundle` (`_maxValSize` in the Alonzo
+-- ledger)
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs]: IIRC, Haddock requires the use of normal single quotes:

```suggestion
-- | The maximum size of a serialized 'TokenBundle'.
-- ('_maxValSize' in the Alonzo ledger)
```

File: lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    it "unit_assessTokenBundleSize_fixedSizeBundle_32" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_32
+    it "unit_assessTokenBundleSize_fixedSizeBundle_48" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_48
+    it "unit_assessTokenBundleSize_fixedSizeBundle_64" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_64
+    it "unit_assessTokenBundleSize_fixedSizeBundle_128" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_128
+
+
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs]:
```suggestion

```

File: lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+------------------------------------------------------------------
+-- Assessing the sizes of token bundles
+--------------------------------------------------------------------------------
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs]:
```suggestion
--------------------------------------------------------------------------------
-- Assessing the sizes of token bundles
--------------------------------------------------------------------------------
```

File: lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs [hunk 9]
```diff
@@ -0,0 +1,224 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Cardano.Wallet.Write.Tx.Balance.TokenBundleSizeAssessorSpec where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessorSpec.hs]: Import in wrong place ^^^

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 10]
```diff
@@ -4108,3 +4111,4 @@ dummyPolicyK = KeyHash Policy (BS.replicate 32 0)
 dummyMnemonic :: SomeMnemonic
 dummyMnemonic = SomeMnemonic $ either (error . show) id
     (entropyToMnemonic @12 <$> mkEntropy "0000000000000000")
+
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Spurious extra line added to end of file.
```suggestion
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 11]
```diff
@@ -3922,6 +3905,26 @@ paymentPartialTx txouts = PartialTx (Cardano.Tx body []) mempty []
...
+        )
+        []
+        Cardano.TxBodyNoScriptData
+        Nothing
+        Cardano.TxScriptValidityNone
+
+    dummyHash = Hash $ B8.replicate 32 '8'
+
+    i = TxIn dummyHash 0
+    utxo = Compatibility.toCardanoUTxO ShelleyBasedEraBabbage $ UTxO $ Map.singleton i (TxOut dummyAddr bundle)
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Line length exceeds the limit.

```suggestion
    utxo = Compatibility.toCardanoUTxO ShelleyBasedEraBabbage
        $ UTxO
        $ Map.singleton i (TxOut dummyAddr bundle)
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 12]
```diff
@@ -3189,8 +3172,8 @@ balanceTransactionGoldenSpec = describe "balance goldens" $ do
             Cardano.toShelleyCertificate
                 <$> mkDelegationCertificates delegationAction (Left xpub)
           where
-            poolId = PoolId "\236(\243=\203\230\214@\n\RS^3\155\208d|\
-                            \\ts\202l\f\249\194\187\230\131\141\198"
+            poolId = PoolId "\236(\243=\203\230\214@\n\RS^3\155\208d|                            \\ts\202l\f\249\194\187\230\131\141\198"
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Is this change intended?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 13]
```diff
@@ -2201,8 +2184,8 @@ distributeSurplusSpec = do
                 sizeOfCoin (Coin $ 2 `power` 32    ) `shouldBe` TxSize 9
 
     describe "costOfIncreasingCoin" $ do
-        it "costs 176 lovelace to increase 4294.967295 ada (2^32 - 1 lovelace) \
-           \by 1 lovelace on mainnet" $ do
+        it "costs 176 lovelace to increase 4294.967295 ada (2^32 - 1 lovelace)            \by 1 lovelace on mainnet" $ do
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Is this change intended?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 14]
```diff
@@ -2121,6 +2094,14 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
                 Left (ErrBalanceTxAssignRedeemers
                         (ErrAssignRedeemersTargetNotFound faultyRedeemer))
   where
+    outputs
+        :: Cardano.Tx Cardano.BabbageEra
+        -> [Babbage.BabbageTxOut StandardBabbage]
+    outputs
+        (Cardano.Tx
+            (Cardano.ShelleyTxBody _ body _ _ _ _ )
+        _) = Write.outputs RecentEraBabbage body
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: Hmm. Couldn't we just write:

```suggestion
    outputs (Cardano.Tx (Cardano.ShelleyTxBody _ body _ _ _ _ ) _) =
        Write.outputs RecentEraBabbage body
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 15]
```diff
@@ -2036,6 +1973,42 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
...
+                    (inputPartialTx (TokenBundle.fromCoin (Coin 1_000_000)))
+
+            let costOfIncreasingCoin' = 4
+            let expectedOuts =
+                    if room - costOfIncreasingCoin' >= 0
+                    then 1
+                    else 2 -- change doesn't fit in a single output
+            case res of
+                Right tx -> length (outputs tx) === expectedOuts
+                Left e -> counterexample (show e) $ property False
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: @Anviking what's the motivation for adding this test?

To check that change bundles are split correctly, we already have a comprehensive set of tests within `Cardano.CoinSelection`. Is there any benefit to repeating this test coverage here?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 16]
```diff
@@ -31,6 +31,7 @@
 {-# OPTIONS_GHC -fno-warn-ambiguous-fields #-}
 #endif
 
+
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]:
```suggestion
```

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    ( NFData )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+import Test.QuickCheck
+    ( Arbitrary (..), oneof )
+
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeAssessor.hs]:
```suggestion
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 15 above)
Comment (Anviking) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]: I added it for additional e2e coverage when I was considering not keeping the tests now in `TokenBundleSizeSpec`.

I will remove it now, but I have a feeling that tests like this could be very important in the future if we are able to use the ledger rules as ground truth.

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize.hs [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+import qualified Data.ByteString as BS
+
+-- | The maximum size of a serialized 'TokenBundle'.
+-- ('_maxValSize' in the Alonzo ledger)
+newtype TokenBundleMaxSize = TokenBundleMaxSize
+    { unTokenBundleMaxSize :: TxSize }
+    deriving (Eq, Generic, Show)
+
+instance NFData TokenBundleMaxSize
+-- | Assesses a token bundle size in relation to the maximum size that can be
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize.hs:50]:
```suggestion

-- | Assesses a token bundle size in relation to the maximum size that can be
```

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize/Gen.hs [hunk 19]
```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE TypeApplications #-}
+module Cardano.Wallet.Write.Tx.Balance.TokenBundleSize.Gen where
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize/Gen.hs]:
```suggestion
module Cardano.Wallet.Write.Tx.Balance.TokenBundleSize.Gen
    ( genTokenBundleMaxSize
    , shrinkTokenBundleMaxSize
    )
    where
```

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs [hunk 20]
```diff
@@ -841,10 +839,9 @@ selectAssets era (ProtocolParameters pp) utxoAssumptions outs redeemers
   where
     selectionConstraints = SelectionConstraints
         { assessTokenBundleSize =
-            withConstraints era $
                 -- TODO (ADP-2967): avoid importing Compatibility.
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs]:
```suggestion
```

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 21]
```diff
@@ -2145,12 +2078,14 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
         (dummyTimeTranslationWithHorizon horizon)
         testStdGenSeed
 
-    utxo coins = UTxO $ Map.fromList $ zip ins outs
+    utxoWithBundles bundles = UTxO $ Map.fromList $ zip ins outs
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:2081]: Slight restructuring of helper functions not needed anymore, but I think it's fine to keep.

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize/Gen.hs [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+genTokenBundleMaxSize :: Gen TokenBundleMaxSize
+genTokenBundleMaxSize = TokenBundleMaxSize . TxSize <$>
+    oneof
+      -- Generate values close to the mainnet value of 4000 (and guard
+      -- against underflow)
+      [ fromIntegral . max 0 . (4000 +) <$> arbitrary @Int
+
+      -- Generate more extreme values (both small and large)
+      , fromIntegral <$> arbitrary @Word64
+      ]
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize/Gen.hs]:
```suggestion
    oneof
        -- Generate values close to the mainnet value of 4000 (and guard
        -- against underflow)
        [ fromIntegral . max 0 . (4000 +) <$> arbitrary @Int
        
        -- Generate more extreme values (both small and large)
        , fromIntegral <$> arbitrary @Word64
        ]
```

File: lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeSpec.hs [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+        property prop_assessTokenBundleSize_shrink
+    it "unit_assessTokenBundleSize_fixedSizeBundle_32" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_32
+    it "unit_assessTokenBundleSize_fixedSizeBundle_48" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_48
+    it "unit_assessTokenBundleSize_fixedSizeBundle_64" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_64
+    it "unit_assessTokenBundleSize_fixedSizeBundle_128" $
+        property unit_assessTokenBundleSize_fixedSizeBundle_128
+--------------------------------------------------------------------------------
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Write/Tx/Balance/TokenBundleSizeSpec.hs]:
```suggestion
        property unit_assessTokenBundleSize_fixedSizeBundle_128

--------------------------------------------------------------------------------
```

**Review by jonathanknowles — APPROVED** [review]
Hi @Anviking 

LGTM!

Just a couple of minor formatting nitpicks. (See comments.)
